File: binarySearch.py
#binary search
#time complexity -> O(logn) used binary +while loop
#(i.e. of binary 16 -> 8 -> 4 -> 2 -> 1 worse case)
#space complexity -> O(1) only used a few extra variables like l, r, and m
class Solution:
    def binarySearch(self, nums: List[int], target: int):
        l = 0
        r = len(nums) - 1

        while l <= r:
            m = (l + r)// 2
            if nums[m] < target:
                l = m + 1
            elif nums[m] > target:
                r = m - 1
            else: #we have found our target
                return m
        return -1

File: binarySearch.cpp
//binary search
//time complexity -> O(log n) because of binary search
//i.e worst case 12 -> 6 -> 3 -> 1
//space complexity -> O(1) in memory because only used constant operations by variables
class Solution{
public:
    int binarySearch(vector<int> nums, int target){
        l = 0;
        r = nums.size() - 1;

        while (l <= r){
            //m = (l + r) / 2; //this might cause an overflow 2^32 if nums is large
            m = l - ((r - l) / 2) //use this to solve overflow
            if(nums[m] < target){
                l = m + 1;
            }
            else if(nums[m] > target){
                r = m - 1;
            }
            else{ //we found our solution
                return m;
            }
        }
        return -1;
    }
};